### PR TITLE
Observe state of each trackable channel

### PR DESCRIPTION
**Pull request by KacperKluka**
I've added a listener for the Ably channel connection state changes. Then I've used it in the calculation of the asset state in the `CorePublisher`.

### CODE REVIEW

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/Ably.kt [hunk 1]
```diff
@@ -139,6 +153,7 @@ internal class DefaultAbly(
     override fun disconnect(trackableId: String, presenceData: PresenceData, callback: (Result<Unit>) -> Unit) {
         val removedChannel = channels.remove(trackableId)
         if (removedChannel != null) {
+            removedChannel.unsubscribe()
```
Comment (QuintinWillison) [inline publishing-sdk/src/main/java/com/ably/tracking/publisher/Ably.kt:156]: This change looks unrelated to the purpose of this pull request? Is it a (related) bug fix?

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/CorePublisher.kt [hunk 2]
```diff
@@ -334,6 +346,10 @@ constructor(
         }
     }
 
+    private fun getLastChannelConnectionStateChange(state: PublisherState, trackableId: String): ConnectionStateChange =
+        state.lastChannelConnectionStateChanges[trackableId]
+            ?: ConnectionStateChange(ConnectionState.OFFLINE, ConnectionState.OFFLINE, null)
```
Comment (QuintinWillison) [inline publishing-sdk/src/main/java/com/ably/tracking/publisher/CorePublisher.kt:351]: It looks like you're creating a new instance every time.
Could this be declared as a constant singleton on `ConnectionStateChange`?

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/CorePublisher.kt [hunk 3]
```diff
@@ -316,15 +317,26 @@ constructor(
...
     private fun updateAssetStatuses(state: PublisherState, trackableId: String) {
         val hasSentAtLeastOneLocation: Boolean = state.lastSentEnhancedLocations[trackableId] != null
+        val lastChannelConnectionStateChange = getLastChannelConnectionStateChange(state, trackableId)
         val newStatus = when (state.lastConnectionStateChange.state) {
-            ConnectionState.ONLINE -> if (hasSentAtLeastOneLocation) AssetStatus.Online() else AssetStatus.Offline()
+            ConnectionState.ONLINE -> {
+                when (lastChannelConnectionStateChange.state) {
+                    ConnectionState.ONLINE -> if (hasSentAtLeastOneLocation) AssetStatus.Online() else AssetStatus.Offline()
+                    ConnectionState.OFFLINE -> AssetStatus.Offline()
+                    ConnectionState.FAILED -> AssetStatus.Failed(lastChannelConnectionStateChange.errorInformation!!) // are we sure error information will always be present?
```
Comment (QuintinWillison) [inline publishing-sdk/src/main/java/com/ably/tracking/publisher/CorePublisher.kt:337]: What does the comment (_way_) on the end of this line refer to?

**Review by QuintinWillison — APPROVED** [review]

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/Ably.kt (hunk 1 above)
Comment (KacperKluka) [reply publishing-sdk/src/main/java/com/ably/tracking/publisher/Ably.kt:156]: Yes, it's probably an unrelated bug fix :wink:

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/CorePublisher.kt (hunk 2 above)
Comment (KacperKluka) [reply publishing-sdk/src/main/java/com/ably/tracking/publisher/CorePublisher.kt:351]: Issue created https://github.com/ably/ably-asset-tracking-android/issues/216

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/CorePublisher.kt (hunk 3 above)
Comment (KacperKluka) [reply publishing-sdk/src/main/java/com/ably/tracking/publisher/CorePublisher.kt:337]: I'm just wondering if the `errorInformation` will always be present for the `ConnectionState.FAILED` state, because if not the `!!` operator will throw NPE.
